src/ann_solo/reader.py

```diff
@@ -511,3 +966,162 @@ def read_mztab_ssms(filename: str) -> pd.DataFrame:
+    """
+    annotation_mapping = {}
+    sequence_len = len(sequence_mapping)
+    for i in range(1, sequence_len):
+        old_p, new_p = i - 1, sequence_mapping[i - 1]
+        if new_p == sequence_len - 1:
+            annotation_mapping[f"b{i}"] = None
+            annotation_mapping[f"y{sequence_len-i}"] = None
+        elif old_p == new_p:
+            annotation_mapping[f"b{i}"] = f"b{i}"
+            annotation_mapping[f"y{sequence_len-i}"] = f"y{sequence_len-i}"
+        else:
+            annotation_mapping[f"b{i}"] = f"b{new_p + 1}"
+            annotation_mapping[
+                f"y{sequence_len-i}"] = f"y{sequence_len-(new_p+1)}"
+    return annotation_mapping
+
+
+def shuffle_and_reposition(spectrum: MsmsSpectrum) -> MsmsSpectrum:
+    """
+    Creates a decoy spectrum from a real spectrum.
+
+    Parameters
+    ----------
+    spectrum: MsmsSpectrum
+        Real spectrum.
+
+    Returns
+    -------
+    MsmsSpectrum
+        Decoy spectrum.
+    """
+    # annotate original spectrum
+    spectrum.annotate_proforma(spectrum.peptide, 10, "ppm", "by")
+    # parse original spectrum
+    parsed_sequence = proforma.parse(spectrum.proforma)
+    shuffled_sequence, shuffled_sequence_mapping = _shuffle(
+        parsed_sequence[0].sequence)
+    # Get the fragment annotations mapping dictionary
+    shuffled_annotation_mapping = _shuffle_annotation_mapping(
+        shuffled_sequence_mapping)
+    # Constract decoy proteoform
+    for modification in parsed_sequence[0].modifications:
+        setattr(modification, 'position',
+                shuffled_sequence_mapping[modification.position])
+    decoy_proforma = proforma.Proteoform(shuffled_sequence,
+                                         parsed_sequence[0].modifications)
+
+    # Get theoretical fragment M/Z of the shuffled sequence
+    decoy_theoretical_fragments = {str(ion_type): mz for ion_type, mz in
+                                   fragment_annotation.get_theoretical_fragments(
+                                       decoy_proforma)}
+    # Reposition peaks
+    mz_shuffled, intensity_shuffled, annotation_shuffled = np.zeros_like(
+        spectrum.mz), np.zeros_like(spectrum.intensity), np.full_like(
+        spectrum.annotation, None, object)
+    for i in range(len(spectrum.mz)):
+        # Peak not annotated. Keep it at the same position.
+        if str(spectrum.annotation[i]) == '?':
+            intensity_shuffled[i] = spectrum.intensity[i]
+            mz_shuffled[i] = spectrum.mz[i]
+        # Peak annotated. Move it to the new position.
+        else:
+            intensity_shuffled[i] = spectrum.intensity[i]
+            annotation_shuffled[i] = shuffled_annotation_mapping[
+                spectrum.annotation[i][0].ion_type]
+            mz_shuffled[i] = decoy_theoretical_fragments[
+                annotation_shuffled[i]]
+    # Reorder the arrays based on the sorted mz array
+    reordered_arrays = zip(
```

Comment:
**todo:** Can be done more efficiently with [numpy.argsort](https://numpy.org/doc/stable/reference/generated/numpy.argsort.html).